### PR TITLE
Add get_correction_from_cmt to corrections_services.py

### PR DESCRIPTION
**Pull request by ahiguera-mx**
Before you submit this PR: make sure to put all operations-related information in a wiki-note, a PR should be about code and is publicly accessible

**What is the problem / what does the code in this PR do**
Add get_correction_from_cmt, this function would return any correction value, it replace get_elife and serves as a general purpose function for single values
**Can you briefly describe how it works?**
See https://github.com/XENONnT/corrections/blob/master/CMT/examples/CMT_user_guide.ipynb block [5]
`elife_file = ('elife','ONLINE',True)`
`straxen.get_correction_from_cmt('010126', elife_file)`
**Can you give a minimal working example (or illustrate with a figure)?**
see above
Please include the following if applicable:
  - Update the docstring(s)
  - Update the documentation
  - Tests to check the (new) code is working as desired.
  - Does it solve one of the open issues on github?

Please make sure that all automated tests have passed before asking for a review (you can save the PR as a draft otherwise).

### CODE REVIEW

File: straxen/corrections_services.py [hunk 1]
```diff
@@ -75,8 +77,8 @@ def get_corrections_config(self, run_id, config_model=None):
 
         if 'to_pe_model' in model_type:
             return self.get_pmt_gains(run_id, model_type, global_version)
-        elif 'elife' in model_type:
-            return self.get_elife(run_id, model_type, global_version)
+        elif model_type in single_value_corrections:
+            return self._get_correction(run_id, model_type, global_version)
         elif model_type in corrections_w_file:
             return self.get_config_from_cmt(run_id, model_type, global_version)
         else:
```
Comment (WenzDaniel) [inline straxen/corrections_services.py:84]: Not part of this PR, but since it just strikes my eye. Can you also add to the raise "... following values are supproted: ...." This often helps to spot the mistake.

File: straxen/plugins/event_processing.py [hunk 2]
```diff
@@ -375,7 +375,7 @@ def compute(self, events):
             (170118_1327, pax_file('XENON1T_s2_xy_ly_SR1_v2.2.json'))]),
    strax.Option(
         'elife_file',
-        default=aux_repo + '3548132b55f81a43654dba5141366041e1daaf01/strax_files/elife.npy',
+        default=('elife_constant', 700000., True),
```
Comment (WenzDaniel) [inline straxen/plugins/event_processing.py:378]: According to the help the number should be a string.

File: straxen/plugins/nveto_recorder.py [hunk 3]
```diff
@@ -16,8 +16,8 @@
                  help="Pretrigger time before coincidence window in ns."),
     strax.Option('resolving_time_recorder_nv', type=int, default=600,
                  help="Resolving time of the coincidence in ns."),
-    strax.Option('nbaseline_samples_lone_records_nv', type=int,
-                 default_by_run=[(0, 10), (12684, 26)], track=True,
+    strax.Option('nbaseline_samples_lone_records_nv', type=tuple,
+                 default=('baseline_samples_nv', 'ONLINE', True), track=True,
```
Comment (WenzDaniel) [inline straxen/plugins/nveto_recorder.py:20]: Thanks for changing. 👍

File: straxen/plugins/nveto_recorder.py [hunk 4]
```diff
@@ -114,8 +114,9 @@ def compute(self, raw_records_nv, start, end):
 
         lr = strax.sort_by_time(lr)
         strax.zero_out_of_bounds(lr)
+        baseline_samples = straxen.get_correction_from_cmt(self.run_id, self.config['nbaseline_samples_lone_records_nv'])
```
Comment (WenzDaniel) [inline straxen/plugins/nveto_recorder.py:117]: Could you please move this line into a
```
def setup(self):
    ....
```
Like n corrected areas, if you do not have time no problem then I change it.

File: straxen/plugins/veto_pulse_processing.py [hunk 5]
```diff
@@ -64,8 +63,9 @@ def compute(self, raw_records_coin_nv):
 
         r = strax.sort_by_time(r)
         strax.zero_out_of_bounds(r)
+        baseline_samples = straxen.get_correction_from_cmt(self.run_id, self.config['baseline_samples_nv'])
```
Comment (WenzDaniel) [inline straxen/plugins/veto_pulse_processing.py:66]: Same here.

File: tests/test_plugins.py [hunk 6]
```diff
@@ -21,8 +21,10 @@
     gain_model=
     ('to_pe_per_run',
      straxen.aux_repo + '58e615f99a4a6b15e97b12951c510de91ce06045/fax_files/to_pe_nt.npy'),
-    s2_xy_correction_map=straxen.pax_file('XENON1T_s2_xy_ly_SR0_24Feb2017.json'),
-    elife_file=straxen.aux_repo + '3548132b55f81a43654dba5141366041e1daaf01/strax_files/elife.npy'
+    s2_xy_correction_map = straxen.pax_file('XENON1T_s2_xy_ly_SR0_24Feb2017.json'),
+    elife_file = ('elife_constant', 700000., True),
+    baseline_samples_nv = ('baseline_samples_nv_constant', 10., False),
+    nbaseline_samples_lone_records_nv = ('baseline_samples_nv_constant', 10., False)
```
Comment (WenzDaniel) [inline tests/test_plugins.py:27]: Why do you set here is_nT == False?

File: straxen/get_corrections.py [hunk 7]
```diff
@@ -83,32 +83,33 @@ def get_to_pe(run_id, gain_model, n_pmts):
...
-        warn("get_elife will be replaced by CorrectionsManagementSevices",
-             DeprecationWarning, 2)
-        # Let's remove these functions and only rely on the CMT in the future
-        x = straxen.get_resource(elife_conf, fmt='npy')
-        run_index = np.where(x['run_id'] == int(run_id))[0]
-        if not len(run_index):
-            # Electron lifetime not known: using placeholders
-            e = 623e3
+def get_correction_from_cmt(run_id, conf):
+    if isinstance(conf, tuple) and len(conf) == 3:
```
Comment (WenzDaniel) [inline straxen/get_corrections.py:87]: Since you have no else raise case here (due to the nested structure I guess), maybe it is easier to negate the statement `if not ... raise ValueError('Expected ....')`

File: straxen/get_corrections.py [hunk 8]
```diff
@@ -83,32 +83,33 @@ def get_to_pe(run_id, gain_model, n_pmts):
...
+        is_nt = conf[-1]
+
+        # check you are not asking for a cte value....
+        model_type, global_version = conf[:2]
+        if 'constant' in model_type:
+            if not isinstance(global_version, float):
+                raise ValueError(f'User specify a model type {model_type} '
+                                 f'and should provide a number. Got: '
+                                 f'{type(global_version)}')
```
Comment (WenzDaniel) [inline straxen/get_corrections.py:96]: Would you not need a similar test also in the else case (of if constant)?

File: straxen/corrections_services.py [hunk 9]
```diff
@@ -75,8 +77,8 @@ def get_corrections_config(self, run_id, config_model=None):
 
         if 'to_pe_model' in model_type:
```
Comment (WenzDaniel) [inline straxen/corrections_services.py:78]: Is not `model_type` just a string? If yes we should change it in `==` as `"to_pe_model" in "to_pe_model_bla"` would be evaluate true too.

File: straxen/get_corrections.py [hunk 10]
```diff
@@ -83,32 +83,33 @@ def get_to_pe(run_id, gain_model, n_pmts):
...
+            if 'samples' in model_type:  # samples are int others are float 
+                return int(global_version)
+            else:
+                return float(global_version)
```
Comment (WenzDaniel) [inline straxen/get_corrections.py:100]: Duplication with 109-112. Could you not rename "global_version" and/or "correction" and have this if-else case only once at the end?

**Review by WenzDaniel — COMMENTED** [review]
Hi Aaron, thank you very much for your nice and clear PR. I just added some smaller comments and questions.

File: straxen/corrections_services.py (hunk 1 above)
Comment (ahiguera-mx) [reply straxen/corrections_services.py:84]: done

File: straxen/plugins/event_processing.py (hunk 2 above)
Comment (ahiguera-mx) [reply straxen/plugins/event_processing.py:378]: update the help accordingly

File: straxen/plugins/nveto_recorder.py (hunk 4 above)
Comment (ahiguera-mx) [reply straxen/plugins/nveto_recorder.py:117]: done

File: straxen/plugins/veto_pulse_processing.py (hunk 5 above)
Comment (ahiguera-mx) [reply straxen/plugins/veto_pulse_processing.py:66]: done

File: tests/test_plugins.py (hunk 6 above)
Comment (ahiguera-mx) [reply tests/test_plugins.py:27]: fixed typo

File: straxen/get_corrections.py (hunk 7 above)
Comment (ahiguera-mx) [reply straxen/get_corrections.py:87]: add else to raise ValueError

File: straxen/get_corrections.py (hunk 8 above)
Comment (ahiguera-mx) [reply straxen/get_corrections.py:96]: Done

File: straxen/get_corrections.py (hunk 10 above)
Comment (ahiguera-mx) [reply straxen/get_corrections.py:100]: done

File: straxen/corrections_services.py (hunk 9 above)
Comment (ahiguera-mx) [reply straxen/corrections_services.py:78]: no, since we have different detectors, see the following lines 

``` if 'to_pe_model' in model_type:                                                                                        
       # Get the detector name based on the requested model_type                                                          
       # This also will be used to the cachable name convention                                                           
       # pmt == TPC, n_veto == n_veto's PMT, etc                                                                          
       detector_names = {'to_pe_model': 'pmt',                                                                            
                                         'to_pe_model_nv': 'n_veto',                                                                      
                                         'to_pe_model_mv': 'mu_veto'}                                                                     
       target_detector = detector_names[model_type]```

**Comment by ahiguera-mx** [comment]
Thanks a lot for very useful comments @WenzDaniel, I've implemented let me know if you have missed something

File: straxen/plugins/veto_pulse_processing.py [hunk 11]
```diff
@@ -50,6 +49,11 @@ class nVETOPulseProcessing(strax.Plugin):
     data_kind = 'records_nv'
     ends_with = '_nv'
 
+    def setup(self):
+
+        self.baseline_samples = straxen.get_correction_from_cmt(self.run_id,
+                                                           self.config['baseline_samples_nv'])
+
```
Comment (WenzDaniel) [inline straxen/plugins/veto_pulse_processing.py:56]: Regarding the failed test, muVETOPulseProcessing is a child of nVETOPulsePorcessing which overwrite the setting for `baseline_samples_nv` with an integer. Hence you have to add [here](https://github.com/XENONnT/straxen/blob/8ae2886c5e68503775782ccea38ba25f9ae54cd6/straxen/plugins/veto_pulse_processing.py#L250) a setup function too which overwrites the setup of nVETOPulseProcessing. The code would simply be:

```
def setup(self):
    self.baseline_samples = self.config['baseline_samples_mv']
```

File: straxen/corrections_services.py [hunk 12]
```diff
@@ -13,6 +13,8 @@
 corrections_w_file = ['mlp_model', 'gcn_model', 'cnn_model',
                       's2_xy_map', 's1_xy_map', 'fdc_map']
 
+single_value_corrections =['elife', 'baseline_samples_nv',
+                           'electron_drift_velocity']
```
Comment (jorana) [inline straxen/corrections_services.py:17]: Perhaps we want to at some point query the database for this info

File: straxen/get_corrections.py [hunk 13]
```diff
@@ -83,33 +83,33 @@ def get_to_pe(run_id, gain_model, n_pmts):
...
+    if isinstance(conf, tuple) and len(conf) == 3:
+        is_nt = conf[-1]
+
+        model_type, global_version = conf[:2]
+        correction = global_version  # in case is a single value
+        if 'constant' in model_type:
+            if not isinstance(global_version, (float, int)):
+                raise ValueError(f'User specify a model type {model_type} '
+                                 f'and should provide a number. Got: '
+                                 f'{type(global_version)}')
```
Comment (jorana) [inline straxen/get_corrections.py:96]: We don't use this for the models like mlp etc.right, perhaps more consistent to also not do it here

File: straxen/plugins/event_processing.py [hunk 14]
```diff
@@ -374,13 +374,11 @@ def compute(self, events):
             (0, pax_file('XENON1T_s2_xy_ly_SR0_24Feb2017.json')),
             (170118_1327, pax_file('XENON1T_s2_xy_ly_SR1_v2.2.json'))]),
    strax.Option(
-        'elife_file',
-        default=aux_repo + '3548132b55f81a43654dba5141366041e1daaf01/strax_files/elife.npy',
-        help='Electron lifetime model '
-             'To use a constant value, provide a link (as this default) To use'
-             'the corrections management tools specify the following:'
+        'elife_conf',
+        default=('elife_constant', 700000., True),
```
Comment (jorana) [inline straxen/plugins/event_processing.py:378]: Can you revert the default for 1T in the context of 1t?

File: straxen/plugins/nveto_recorder.py (hunk 3 above)
Comment (jorana) [inline straxen/plugins/nveto_recorder.py:20]: Would this not also be ("CMT_model", ('baseline_samples_nv', "ONLINE"), True)

File: straxen/plugins/nveto_recorder.py [hunk 15]
```diff
@@ -56,6 +56,11 @@ class nVETORecorder(strax.Plugin):
 
     data_kind = {key: key for key in provides}
 
+    def setup(self):
+
```
Comment (jorana) [inline straxen/plugins/nveto_recorder.py:60]: @WenzDaniel , would you like to have a line 
```python
if isinstance(self.config['nbaseline_samples_lone_records_nv']), int):
  self.baseline_samples = self.config['nbaseline_samples_lone_records_nv']
else:
  self.baseline_samples = straxen.get_correction_from_cmt(self.run_id,
                                                           self.config['nbaseline_samples_lone_records_nv'])
```

File: straxen/plugins/nveto_recorder.py [hunk 16]
```diff
@@ -16,8 +16,8 @@
                  help="Pretrigger time before coincidence window in ns."),
     strax.Option('resolving_time_recorder_nv', type=int, default=600,
                  help="Resolving time of the coincidence in ns."),
-    strax.Option('nbaseline_samples_lone_records_nv', type=int,
-                 default_by_run=[(0, 10), (12684, 26)], track=True,
+    strax.Option('nbaseline_samples_lone_records_nv', type=tuple,
```
Comment (jorana) [inline straxen/plugins/nveto_recorder.py:19]: @WenzDaniel should we not change this option to be the same as in the veto_pulse_processing.py: 
```baseline_samples_nv``` 

Now there are two names for the same thing

File: straxen/plugins/veto_pulse_processing.py [hunk 17]
```diff
@@ -50,6 +49,11 @@ class nVETOPulseProcessing(strax.Plugin):
     data_kind = 'records_nv'
     ends_with = '_nv'
 
+    def setup(self):
+
```
Comment (jorana) [inline straxen/plugins/veto_pulse_processing.py:53]: Same question as above @WenzDaniel

**Review by jorana — COMMENTED** [review]
Thanks Aaron, a few questions also in the light of our recent chat

File: straxen/plugins/nveto_recorder.py (hunk 16 above)
Comment (WenzDaniel) [reply straxen/plugins/nveto_recorder.py:19]: Yes, true I forgot to mention this. Would be nice if you could also this into baseline_samples_nv Aaron. Thanks a lot.

File: straxen/plugins/nveto_recorder.py (hunk 15 above)
Comment (WenzDaniel) [reply straxen/plugins/nveto_recorder.py:60]: This solution would work too and is maybe a bit more flexible. Then you also do not have to add a new setup to muVETOProcessing.

File: straxen/contexts.py [hunk 18]
```diff
@@ -46,7 +46,7 @@
     # Clustering/classification parameters
     s1_max_rise_time=100,
     s2_xy_correction_map=("CMT_model", ('s2_xy_map', "ONLINE"), True),
-    elife_file=("elife_model", "ONLINE",True),
+    elife_conf=("elife", "ONLINE", True),
```
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/contexts.py:49]: It is tricky, since there is not a model in CMT but actual measurement values, so I will leave this discussion for when we revise all strax.Options() to make sure everything is consistent, then we can look at the name conventions of CMT.

File: straxen/plugins/veto_pulse_processing.py (hunk 11 above)
Comment (ahiguera-mx) [reply straxen/plugins/veto_pulse_processing.py:56]: thanks, will add that

File: straxen/corrections_services.py (hunk 12 above)
Comment (ahiguera-mx) [reply straxen/corrections_services.py:17]: sure, we can do that in the future

File: straxen/get_corrections.py (hunk 13 above)
Comment (ahiguera-mx) [reply straxen/get_corrections.py:96]: there is not a constant option for conf with a file

File: straxen/contexts.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (jorana) [reply straxen/contexts.py:49]: The Online version might be, but version1, version2 etc won't be.

File: straxen/plugins/event_processing.py (hunk 14 above)
Comment (ahiguera-mx) [reply straxen/plugins/event_processing.py:378]: sure

File: straxen/plugins/nveto_recorder.py (hunk 3 above)
Comment (ahiguera-mx) [reply straxen/plugins/nveto_recorder.py:20]: see above comment

File: straxen/plugins/nveto_recorder.py (hunk 15 above)
Comment (ahiguera-mx) [reply straxen/plugins/nveto_recorder.py:60]: I would go with the option of adding a `setup()` function for consistency, if you want to change please go ahead and implement it.

File: straxen/plugins/veto_pulse_processing.py (hunk 17 above)
Comment (ahiguera-mx) [reply straxen/plugins/veto_pulse_processing.py:53]: see above comment

File: straxen/contexts.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ahiguera-mx) [reply straxen/contexts.py:49]: we don't know, they might come from Kr83 measurement

File: straxen/plugins/nveto_recorder.py (hunk 16 above)
Comment (ahiguera-mx) [reply straxen/plugins/nveto_recorder.py:19]: done

**Comment by ahiguera-mx** [comment]
Hi @WenzDaniel and @jorana 
Thanks for the useful comments

**Review by WenzDaniel — APPROVED** [review]
Hi Aaron thanks a lot I do not have any further comments. Maybe you want to address the comments from our new code style guard "reviewdog", for a better coding style. But this is optional.

File: straxen/plugins/nveto_recorder.py (hunk 3 above)
Comment (jorana) [reply straxen/plugins/nveto_recorder.py:20]: I still disagree, if we want to make the options to call CMT consistent, we should do it from the start. Especially for such low level things like baseline options. Perhaps you and @tunnell can come up with some consistent naming scheme to interface with CMT because I don't have time I'm afraid.

Comment (ahiguera-mx) [reply straxen/plugins/nveto_recorder.py:20]: Hi @jorana 
I don't think this PR neither GitHub is the forum to discuss this and since you brought it let me clarify that we have this naming convention because you wanted to keep the `to_pe_per_run` available in the first place. If we agree to disable this options we can work towards a simple naming convention for CMT
